src/main/jbake/content/blog/2017/07/opentracing-jaxrs.adoc

```diff
@@ -0,0 +1,136 @@
+It is sufficient for our demo and in the next step we are going to add instrumentation and simple business logic.
+Do not worry at the bottom of this article are all the necessary links including source code.
+
+== Instrumentation
+To add JAX-RS OpenTracing instrumentation is very simple, just include the following dependency on
+the classpath and tracing feature will be automatically registered.
+
+----
+<dependency>
+    <groupId>io.opentracing.contrib</groupId>
+    <artifactId>opentracing-jaxrs2</artifactId>
+</dependency>
+----
+
+OpenTracing is just an API, therefore it is required to register a specific tracer instance. In this demo
+we are going to use https://github.com/uber/jaeger[Jaeger tracing system]. Tracer should be created
+and initialized only once per process, hence `ServletContextListener` is the ideal place for this task:
+
+[source,java]
+----
+@WebListener
+public class TracingContextListener implements ServletContextListener {
+
+  @Inject
+  private io.opentracing.Tracer tracer;
+
+  @Override
+  public void contextInitialized(ServletContextEvent sce) {
+    GlobalTracer.register(jaegerServer());
+  }
+
+  @Override
+  public void contextDestroyed(ServletContextEvent sce) {}
+
+  @Produces
+  @Singleton
+  public static io.opentracing.Tracer jaegerTracer() {
+    return new Configuration("wildfly-swarm", new Configuration.SamplerConfiguration(
+        ProbabilisticSampler.TYPE, 1),
+        new Configuration.ReporterConfiguration())
+        .getTracer();
+  }
+}
+----
+
+As you noticed we are using CDI to share a tracer instance in our app. If we forgot to register
+specific tracer instance then tracing feature would use `NoopTracer`. Now we can verify tracing
+by starting Jaeger server `docker run --rm -it --network=host jaegertracing/all-in-one` and
+hitting hello endpoint. Trace with one span should be present in the UI
+accessible on `http://localhost:16686`.
+
+== Instrumenting business logic
+JAX-RS instrumentation provides out of the box a nice visibility into your app, however, often
+it's necessary to add custom data to the trace to see what is happening in service or database layer.
+
+Following code snippet shows how service layer can create and add data to the trace:
+
+[source,java]
+----
+public class BackendService {
+
+  @Inject
+  private io.opentracing.Tracer tracer;
+
+  public String action() throws InterruptedException {
+    int random = new Random().nextInt(200);
+
+    try (ActiveSpan span = tracer.buildSpan("action").startActive()) {
+      anotherAction();
+      Thread.sleep(random);
+    }
+
+    return String.valueOf(random);
+  }
+
+  private void anotherAction() {
+    tracer.activeSpan().setTag("anotherAction", "data");
+  }
+----
+
+Note that it's not necessary to manually pass a span instance around. Method `anotherAction` access
+current active span from the tracer. If we inject this class into REST endpoint and call `action` method
```

Comment:
Maybe rewrite the "If we inject...." and following sentence with something like:
"With the additional instrumentation shown above, an invocation of the REST endpoint would result in a trace consisting of two spans, one representing the inbound server request, and the other the business logic. The span representing the server processing is ...."